Support hidden chapters

maarut opened the pull request:
Fixes #

Added support for hidden chapters (hidden chapters don't show up in the Table Of Contents).
When scrubbing through the timeline, chapter titles, URLs and images are updated based on the position of the playhead.

## To test

I tested using the Stratechery Podcast episode dated 25th October 2022. It has multiple hidden chapters with images.

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

maarut committed:
Initial support for hidden chapters
- Make use of the "hidden" flag of the chapter data
- Display only visible chapters in chapter lists
- Only perform skip operations on visible chapters
- Correctly display images and links when hidden chapters are being
  played through

maarut committed:
When scrubbing the timeline, the chapter information is also updated.

maarut committed:
Update Now Playing Screen chapter information when scrubbing through the
playback slider.

maarut committed:
Update Changelog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Automattic/pocket-casts-ios?pullRequest=553) <br/>All committers have signed the CLA.

maarut commented:
[This is the first time I've submitted a PR for an open source project. Please advise if I haven't correctly followed your processes. I can close this PR and re-create a new one following the proper process if necessary]

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -0,0 +1,60 @@
+//  Created by Maarut Chandegra on 11/11/2022.
+//  Copyright © 2022 Shifty Jelly. All rights reserved.
+//
+
```
Can you remove the header from this file?

maarut replied:
I've removed all copyright notices.

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -0,0 +1,60 @@
+//  Copyright © 2022 Shifty Jelly. All rights reserved.
+//
+
+import Foundation
```
```suggestion
```

Since UIKit already includes Foundation we don't need to double the imports up

maarut replied:
I've removed superfluous imports

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -0,0 +1,60 @@
+        visibleChapter()?.duration ??
+        chapters.max(by: { $0.duration < $1.duration })?.duration ??
+        1
+    }
```
We can reduce the overall code changes by making these all be read-only computed properties.

maarut replied:
I've reworked this code to be computed properties.

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -0,0 +1,60 @@
+class Chapters: Equatable {
+    private var chapters = [ChapterInfo]()
+
+    init(chapters: [ChapterInfo]) {
```
I see elsewhere in the code you're init'ing with an empty array. I think we can clean up the code by adding a default value, that way you can just do `Chapters()`

```suggestion
    init(chapters: [ChapterInfo] = []) {
```

maarut replied:
Default value defined on init.

emilylaguna reviewed (changes requested):
👋 Hi @maarut, Thank you so much for your contribution! We are super thrilled that you chose to contribute to Pocket Casts for your **first ever** OSS project! 

Since this PR has a lot of changes we will need to do some extra testing on our end so our reviews may be a bit delayed. However I did a quick review and left some comments below. 

In addition to the comments below, I think these changes could benefit from adding some unit tests to make sure everything works as intended.

Let me know what you think.

maarut committed:
Merge remote-tracking branch 'ssh/trunk' into support_hidden_chapters

maarut committed:
Merge remote-tracking branch 'ssh/trunk' into support_hidden_chapters

maarut committed:
- Add tests to new Chapters class
- Verify that chapters overlap when creating Chapters object
- Rework functions on Chapters class to properties for easier use
- Provide default values on Chapters functions for easier use

maarut commented:
@emilylaguna Thanks for providing review comments and suggestions for improvements. Sorry for the delay, but I think I have addressed all outstanding comments. I hope this gets us a step closer to merging the PR.

emilylaguna commented on podcasts/ChapterManager.swift:
```diff
@@ -12,7 +12,7 @@ class ChapterManager {
     private var lastEpisodeUuid = ""
 
-    var currentChapters = Chapters(chapters: [])
+    var currentChapters: Chapters = Chapters()
```
We can remove the Chapters type here since it's inferred from the `Chapters()` init.

```suggestion
    var currentChapters = Chapters()
```

maarut replied:
Removed type specification

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -1,60 +1,83 @@
     }
 }
+
+private func chaptersOverlap(_ chapters: [ChapterInfo]) -> Bool {
```
While this does work, the preferred way to accomplish this would be to use a private extension like:

```suggestion
private extension Chapters {
    func chaptersOverlap(_ chapters: [ChapterInfo]) -> Bool {
```

maarut replied:
Moved code in to a private class extension as suggested.

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -1,60 +1,83 @@
+    }
+
+    init?(chapters: [ChapterInfo]) {
+        if !chaptersOverlap(chapters) {
+            return nil
```
Maybe a dumb question, but could you explain the purpose of doing the chapters overlap check and why we want to return nil if they don't overlap?

maarut replied:
The purpose of the class is to represent the chapters for an audio file at a given playback position. Given this purpose, if the chapters don't overlap over that given playback position, then the query used to extract the chapters isn't valid for the use of this class, and there is a potential for returning incorrect data to users of the class. I think a reasonable case at this point is therefore to return nil.

Given your later comment regarding an initialiser returning nil, I have changed my opinion on returning nil. A Chapters object created with no chapters can operate correctly by users of the class. They can simply check if there are a non-zero number of chapters contained in the object prior to querying it for title, image, url, etc.

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -1,60 +1,83 @@
+
+private func chaptersOverlap(_ chapters: [ChapterInfo]) -> Bool {
+    let ranges = chapters.compactMap { $0.duration > 0 ? $0.startTime.seconds ... ($0.startTime.seconds + $0.duration) : nil }
+    var ranegsOverlap = true
```
Small typo here. 

```suggestion
    var rangesOverlap = true
```

maarut replied:
Typo corrected.

emilylaguna commented on podcasts/Chapters.swift:
```diff
@@ -1,60 +1,83 @@
+    init?(chapters: [ChapterInfo]) {
+        if !chaptersOverlap(chapters) {
+            return nil
+        }
```
Since the `chapters` property already sets itself to an empty array we can remove the need for an optional and just only set the property if the overlap check succeeds like so:

```suggestion
    init(chapters: [ChapterInfo] = []) {
        if chaptersOverlap(chapters) {
            self.chapters = chapters
        }
    }
```

maarut replied:
Adopted suggested code.

emilylaguna reviewed (changes requested):
Hey! Thanks for making the edits, the code is shaping up nicely. 

I left a few more comments below, along with some questions. 

Please let me know what you think.

maarut committed:
Merge remote-tracking branch 'ssh/trunk' into support_hidden_chapters

maarut committed:
Address review comments
- Remove optional init on Chapters for easier use.
- Rename `chapterForTime()` function to `chaptersForTime()` to better
  describe what the function does.
- Conform to coding style for the project.

maarut commented:
Sorry for the delay in submitting a summary comment, but I have addressed the simple code related improvements you suggested, and responded to your query around the Chapters class.

CookieyedCodes commented:
@emilylaguna should this be merged now?

emilylaguna reviewed (approved):
Hey @maarut! I apologize, I didn't see your reply and this slipped off my radar. 

I've retested and all looks good. 

I have [made a PR](https://github.com/Automattic/pocket-casts-ios/pull/839) that merges trunk into the PR and moves the changes in the CHANGELOG to the new version, once that is 🟢 on the CI I'll merge it in. 

Thank you for your contribution and patience.

maarut commented:
Thanks for accepting my contribution! No worries about the delay.

leandroalonso commented:
@maarut JFYI, we're reverting the changes made here for the next release: https://github.com/Automattic/pocket-casts-ios/pull/883

While there isn't a problem per se with the code, we had a feature regression and there are a few reports coming in about the missing chapter information — that used to be available.

What I think we can do is reopen this PR making sure that:

1. Episodes that show chapters still show them even after this change;
2. The "Chapters" tab shows the chapter progress
3. The full-player controller has the UI for navigating through chapters

I am more than happy to test and review these changes if you create a PR. :)
